src/Razor/src/Microsoft.VisualStudio.Editor.Razor/BackgroundParser.cs

```diff
@@ -174,16 +176,35 @@ public void QueueChange(ChangeReference change)
             public WorkParcel GetParcel()
             {
                 EnsureNotOnThread(); // Only the background thread can get a parcel
-                _hasParcel.Wait(_cancelSource.Token);
-                _hasParcel.Reset();
-                lock (_stateLock)
+
+                lock (_disposeLock)
                 {
-                    // Create a cancellation source for this parcel
-                    _currentParcelCancelSource = new CancellationTokenSource();
+                    if (_disposed)
+                    {
+                        return WorkParcel.Empty;
+                    }
+                }
+
+                _hasParcel?.Wait(_cancelSource.Token);
```

Comment:
I don't fully understand the reasoning here. When will this be null? If there is a race between line 184 and this, why is this not under a lock?

Reply:
It's a wait statement, if I locked around it then we could never dispose until we had a parcel 😄. The reason why it can be `null` is because once I exit the lock above you could technically get a Dispose call which would result in this being `null` in which case we want to no-op and go back to checking disposed.